stuff/cron/plagiarism_detector.py

```diff
@@ -0,0 +1,243 @@
+sys.path.insert(
+    0,
+    os.path.join(
+        os.path.dirname(os.path.dirname(os.path.realpath(__file__))), "."))
+import lib.db  # pylint: disable=wrong-import-position
+import lib.logs  # pylint: disable=wrong-import-position
+from copydetect import CopyDetector  # type: ignore
+from mysql.connector import errorcode
+
+class Results(NamedTuple):
+    contest_id: int
+    score_1: int
+    score_2: int
+    submission_id_1: int
+    submissino_id_2: int
+    contents: str
+
+# SQL Queries
+
+CONTESTS_TO_RUN_PLAGIARISM_ON = """ SELECT c.`contest_id`, c.`alias`, c.`problemset_id`
+                                    FROM `Contests` as c
+                                    WHERE
+                                        c.`check_plagiarism` = 1 AND
+                                        c.`finish_time` < NOW() - INTERVAL 20 MINUTE AND
+                                        c.`contest_id` NOT IN
+                                            (SELECT p.`contest_id` 
+                                            FROM `Plagiarisms` as p);
+                                """
+
+GET_CONTEST_SUBMISSION_IDS= """ SELECT c.contest_id, s.submission_id, s.problemset_id,
+                                s.problem_id, s.verdict, s.guid, s.language
+                                FROM Submissions as s 
+                                INNER JOIN Contests c ON c.problemset_id = s.problemset_id 
+                                WHERE c.contest_id = %s;
+                            """
+
+INSERT_INTO_PLAGIARISMS = """  
+                                INSERT INTO `Plagiarisms`
+                                (`contest_id`, `score_1`, `score_2` , `submission_id_1` , `submission_id_2`, `contents`)
+                                VALUES
+                                (%s, %s, %s, %s, %s, %s)
+                            """
+
+#Constants
+START_RED = "<span class='highlight-red'>"
+START_GREEN = "<span class='highlight-green'>"
+END = "</span>"
+
+SubmissionDownloader = Callable[[str, str], None]
+
+class S3SubmissionDownloader:
+    ''' A SubmissionDownloader that can download files from an S3 bucket.  '''
+
+    def __init__(self, bucket_name: str = 'omegaup-backup') -> None:
+        self._bucket = boto3.client('s3').Bucket(bucket_name)
+    
+    def __call__(self, guid: str, destination_path: str) -> None:
+        self._bucket.download_file(f'omegaup/submissions/{guid[:2]}/{guid[2:]}', destination_path)
+    
+class LocalSubmissionDownloader:
+
+    def __init__(self, dir: str) -> None:
+        self._dir = dir
+    
+    def __call__(self, guid: str, destination_path: str) -> None:
+        
+        shutil.copyfile(os.path.join(self._dir, f'submissions/{guid[:2]}/{guid[2:]}'), destination_path)
+
+"""
+    Function to get the range of lines that are plagiarised. 
+    The length of each list should be even. 
+    [1, 4, 7, 9] So, the ranges are 1-4 and 7-9. 
+"""
+
+def get_range(code: Sequence[str]) -> Sequence[int]:
+
+    code_range = []
```

Comment:
```suggestion
    code_range: List[int] = []
```

let's type all local declarations that assign to `[]` or `{}`.

Reply:
Done for all. I also included List in typing module.